## Cleaning

In [119]:
import pandas as pd
df = pd.read_csv("name_data/company_person_name_dataset.csv")
df

name  class lang
0                The Canal of the Angels      0   en
1                      Rescue Renovation      0   en
2       Agatha Christie: The ABC Murders      0   en
3                            Siti Akbari      0   ar
4                                  Stany      0   pl
...                                  ...    ...  ...
199995                   Robber's Bridge      0   en
199996                       Johan Renck      0   en
199997                      Lyle Stewart      1   en
199998           Thomas Colclough Watson      1   en
199999                              Gavà      0   ca

[200000 rows x 3 columns]

In [120]:
df = df[df["class"]==1]
#look into smote for imbalance

In [121]:
import numpy as np

In [122]:
print(np.any(df["name"].isnull()))
print(np.any(df["class"].isnull()))
print(np.any(df["lang"].isnull()))

True
False
False


In [123]:
print(len(df.isnull()))

48190


In [124]:
df = df.dropna()

In [125]:
df.drop_duplicates()

name  class lang
11                 Raha Etemadi      1   et
12       Leena Peltonen-Palotie      1   fi
16                  Luma Grothe      1   en
20                Takuya Kakine      1   ja
21               Ōuyáng Zhènhuá      1   mi
...                         ...    ...  ...
199976          Clyde Donaldson      1   en
199980          Terry Alexander      1   en
199989             Neil Roebuck      1   en
199997             Lyle Stewart      1   en
199998  Thomas Colclough Watson      1   en

[47759 rows x 3 columns]

In [126]:
language_counts = df["lang"].value_counts
print("Language distribution:")
print(language_counts)

Language distribution:
<bound method IndexOpsMixin.value_counts of 11        et
12        fi
16        en
20        ja
21        mi
          ..
199976    en
199980    en
199989    en
199997    en
199998    en
Name: lang, Length: 48188, dtype: object>


In [127]:
print(np.any(df["name"].isnull()))
print(np.any(df["class"].isnull()))
print(np.any(df["lang"].isnull()))
print(np.any(df.duplicated()))

False
False
False
True


In [128]:
df = df.drop_duplicates()

In [129]:
print(np.any(df.duplicated()))

False


In [130]:
print("Language distribution:")
print(language_counts)

Language distribution:
<bound method IndexOpsMixin.value_counts of 11        et
12        fi
16        en
20        ja
21        mi
          ..
199976    en
199980    en
199989    en
199997    en
199998    en
Name: lang, Length: 48188, dtype: object>


In [131]:
df["lang"] = df["lang"].str.strip()
language_counts = df["lang"].value_counts()
print("Language distribution:")
print(language_counts)

Language distribution:
lang
en    23352
es     2532
it     1224
fr     1181
ja     1155
      ...  
my        1
sd        1
kk        1
pa        1
si        1
Name: count, Length: 103, dtype: int64


## Additional Cleaning to Badly Formatted Names


Reset df index starting from 0

In [132]:
df.reset_index(drop=True, inplace=True)
print(df)

                          name  class lang
0                 Raha Etemadi      1   et
1       Leena Peltonen-Palotie      1   fi
2                  Luma Grothe      1   en
3                Takuya Kakine      1   ja
4               Ōuyáng Zhènhuá      1   mi
...                        ...    ...  ...
47754          Clyde Donaldson      1   en
47755          Terry Alexander      1   en
47756             Neil Roebuck      1   en
47757             Lyle Stewart      1   en
47758  Thomas Colclough Watson      1   en

[47759 rows x 3 columns]


Create langname to find the alphabet name for each character in a name instead of just the first letter (Anna said its better to have full list) and create column

In [133]:
import unicodedata

In [134]:
langname = lambda x:[unicodedata.name(char).split(' ')[0] for char in x]
df['alphabet'] = df['name'].apply(langname)
print(df['alphabet'].head())

0    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
1    [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...
2    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
3    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
4    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
Name: alphabet, dtype: object


In [135]:
allalphatype = []
#List of all alphabet types of every character in the dataset
for i in df['name']:
    for x in langname(i):
        allalphatype.append(x)

print(len(set(allalphatype)))

total = 0
for i in set(allalphatype):
    count = 0
    total +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total)

55
418 COMMA
1 ZERO
2 TAMIL
3 ARMENIAN


1 KATAKANA
1 DAGGER
23 HANGUL
5 NUMBER
75 CYRILLIC
2 GEORGIAN
4 KANNADA
1 WHITE
5 MODIFIER
31 COLON
77 ARABIC
13 HEBREW
3 ASTERISK
85 DIGIT
2 HIRAGANA
1 SINHALA
1 LESS-THAN
1 MALAYALAM
1 LOW
1 FULLWIDTH
15 SOLIDUS
5 AMPERSAND
1 EQUALS
5 BENGALI
47392 LATIN
2 IDEOGRAPHIC
1 DOLLAR
10 SEMICOLON
301 LEFT
1 TELUGU
184 CJK
2489 FULL
1 ACUTE
1 MYANMAR
2 THAI
1400 HYPHEN-MINUS
14 GREEK
180 QUOTATION
5 COMBINING
1 FEMININE
339 APOSTROPHE
2 MIDDLE
1 GURMUKHI
1 ETHIOPIC
17 DEVANAGARI
312 RIGHT
4 EN
1 EXCLAMATION
1 SYRIAC
46500 SPACE
4 VERTICAL
55


In [136]:
def findname(character, indexes):
    for x, y in enumerate(df['name']):
        if str(character) in (langname(y)):
            indexes.append(x)
            print(x, y, df.iloc[x, 2])

def replacechar(orig, repl, indexes):
    for y in indexes:
        df.iloc[y, 0]=df.iloc[y, 0].replace(orig, repl)
        print(df.iloc[y, 0])

### Clean ASTERISK

In [137]:
ast = []
findname('ASTERISK', ast)

6816 León Ávalos y Vez* es
12044 Biga*Ranx en
25458 *Tomko sk


In [138]:
tmp = [6816, 25458]
for y in tmp:
    df.iloc[y, 0]=df.iloc[y, 0].replace('*', '')
    print(df.iloc[y, 0])

df.iloc[12044, 0]=df.iloc[12044, 0].replace('*', ' ')
print({df.iloc[i, 0] for i in ast})


León Ávalos y Vez
Tomko
{'Tomko', 'León Ávalos y Vez', 'Biga Ranx'}


### Clean FULLWIDTH

In [139]:
fullw = []
findname('FULLWIDTH', fullw)

43794 Thomas Prence （or，Thomas Prince） en


In [140]:
df.iloc[43794, 0]= "Thomas Prence or Thomas Prince"
print(df.iloc[43794, 0])

Thomas Prence or Thomas Prince


### Clean QUOTATION

In [141]:
quo = []
findname('QUOTATION', quo)

190 Francis "Frank" Forshew sv
199 Evelyn "Champagne" King en
248 Ernesto "Boy" F. Herrera en
440 Graeme "Shirley" Strachan en
557 Augustine "Willie" Dominguez en
775 Brian "Goldbelt" Maxine en
834 Joyce "Fenderella" Irby en
883 Farley "Jackmaster" Funk en
1686 Edgar "Ted" Codd en
1780 Lewis Burwell "Chesty" Puller cy
2551 Ivan Maček "Matija" hr
3020 Billy "Harp" Hamilton en
3094 George "Chappie" Johnson Jr. en
3285 "Professor" Irwin Corey en
3391 Alfredo "El Salsero" Escalera es
3477 Peter Ulysses "Sturgis" Turner en
3716 Mark Edmund "Duke" Bainum en
4987 Robert W. "Bob" Straub en
4997 James "J.T." Taylor en
5503 Arthur "Andy" B. VanGundy Jr. te
5743 David L. "Dave" Pearce en
6793 William "Bill" L. Mack en
6942 John Richard "Jack" Williams en
7024 Ronald Dean "Ron" Givens en
7480 H. K. "Kenneth" Cassidy en
7725 "Şeker" Ahmed Pasha tr
8116 H. L. "Matty" Matthews en
8296 Abigail "Abby" Sciuto it
8436 Hoffman Lee "Hop" Fuller en
8608 Christopher "Chris" A. Brown en
9339 Donald Howard "Do

In [142]:
replacechar('"', '', quo)

Francis Frank Forshew
Evelyn Champagne King
Ernesto Boy F. Herrera
Graeme Shirley Strachan
Augustine Willie Dominguez
Brian Goldbelt Maxine
Joyce Fenderella Irby
Farley Jackmaster Funk
Edgar Ted Codd
Lewis Burwell Chesty Puller
Ivan Maček Matija
Billy Harp Hamilton
George Chappie Johnson Jr.
Professor Irwin Corey
Alfredo El Salsero Escalera
Peter Ulysses Sturgis Turner
Mark Edmund Duke Bainum
Robert W. Bob Straub
James J.T. Taylor
Arthur Andy B. VanGundy Jr.
David L. Dave Pearce
William Bill L. Mack
John Richard Jack Williams
Ronald Dean Ron Givens
H. K. Kenneth Cassidy
Şeker Ahmed Pasha
H. L. Matty Matthews
Abigail Abby Sciuto
Hoffman Lee Hop Fuller
Christopher Chris A. Brown
Donald Howard Don Yarborough
Lee Joseph Jody Amedee, III
Katherine Kathy Webb-Kane
Frank Monkey Foreman
Jacob Kobi Baladev
Dan The Dude Mowrey
Charles Charley Kohlhase
Jeremiah Jerry Alvin Jones
Clifton R. Cliff Ammons
Silvano Chori Shueg
Don Mr. Nice Guy Hamilton
Willie Big Eyes Smith
Elinor Ellie Walker
Daniel 

### Clean SEMICOLON

In [143]:
semi = []
findname('SEMICOLON', semi)


12404 Caesar Lucius Aurelius Commodus Antoninus Augustus (180); la
15848 Shinmen Takezō; Miyamoto Bennosuke; Niten Dōraku; Shinmen Musashi no Kami Fujiwara no Genshin ja
28415 Lele; Ezequiel haw
33844 Michael O;Shea en
35732 Steven K. D&#39;Arcy en
38925 Darna; Princess Vara; Olga sv
43280 Given name: Xuanyuan (Hsuan-yuan; 軒轅) zh-CN
44626 Marcus Aurelius Numerius Numerianus (from birth to elevation to Caesar); la
45209 Flavius Honorius (from birth to accession); en
46661 (198 to 211); en


In [144]:
replacechar(';', '', semi)

Caesar Lucius Aurelius Commodus Antoninus Augustus (180)
Shinmen Takezō Miyamoto Bennosuke Niten Dōraku Shinmen Musashi no Kami Fujiwara no Genshin
Lele Ezequiel
Michael OShea
Steven K. D&#39Arcy
Darna Princess Vara Olga
Given name: Xuanyuan (Hsuan-yuan 軒轅)
Marcus Aurelius Numerius Numerianus (from birth to elevation to Caesar)
Flavius Honorius (from birth to accession)
(198 to 211)


### Clean VERTICAL (dropped irrelevant)

In [145]:
ver = []
findname('VERTICAL', ver)

2816 thumb|right|Spitfire Mk IIa P7350 of the BBMF is the only existing airworthy Spitfire that fought in the Battle of Britain. en
4319 frameless|center en


14690 Veronika Sramatythumb|Veronika Sramaty making of The Top Ten en
22302 Timmy O'Dowdthumb|Timmy Dowd in Dowdies December 2011 en


In [146]:
drop = [2816, 4319]
df.iloc[14690, 0] = 'Veronika Sramaty'
df.iloc[22302, 0] = "Timmy O'Dowd"

df['name'][ver]

2816     thumb|right|Spitfire Mk IIa P7350 of the BBMF ...
4319                                      frameless|center
14690                                     Veronika Sramaty
22302                                         Timmy O'Dowd
Name: name, dtype: object

### Clean COLON (incomp) (considering dropping all names)

In [147]:
col = []
findname('COLON', col)

762 Ancestral name: Jiang (姜) gu
3596 German:  Josef Franz Leopold en
4092 German: Henriette Maria Norberta en
5305 Ardeshir (or in higher Persian: Ardashir) Kamkar en
6336 Lily Munster/ OTS name: Phoebe Munster en
7433 File:Hollidaycamo.jpg en
15053 Farid Aslani (Persian : فرید اصلانی) fa
18358 Family name: Hóu (侯) en
19400 Dutch: Gabriël Boudewijn Karel Maria nl
20732 Dutch: Albert Leopold Clemens Maria Meinrad de
21742 Sajid AliUrdu:ساجد  علی fa
22198 Given name: Unknown en
24154 (Arabic:عبد العزيز بديع صالح السليمي) ar
25809 Chinese: Duàn Sīpíng (段思平) zh-CN
25879 Given name: Dao 導 ja
27432 Given name: Zhoupu (州蒲) en
27498 Makhdoom Khusro Bakhtyar (Seraiki:مخدوم  کسر  بختٻد) fa
27503 Laqab: Qutb ad-Din (shortly) uz
27582 Spanish: María Luisa de Borbón y Sajonia es
28808 Maganti Murali Mohan ( Born : Maganti Raja Babu ) te
31311 Traditional Chinese:韓世忠 zh-CN
34236 Given name: Dawud en
37179 Nu:Tone ro
40523 German: Wilhelm Friedrich Karl Ernst Joachim Albrecht en
40992 Family name:Si

### Clean SOLIDUS

In [148]:
sol = []
findname('SOLIDUS', sol)

6336 Lily Munster/ OTS name: Phoebe Munster en
7610 Paula Baracho / Paula Ribeiro pt
8389 Rajendra Choudhary / राजेन्द्रा चौधरी hi
10940 Amanda Porter/Emily Thorne en
12453 Anne-Marie Baiynd/Presnell/Buttrick/Band en
15445 Líu Hóngdù (劉弘度) / Líu Bīn (劉玢) zh-CN
21785 Elavarasan a/l Elangowan ms
22295 Gay Stanhope Falcon/Gay Laurence/Tom Laurence/Michael Mike Waring en
23549 Carsten Bo Eriksen / MBD73 da
29724 Marcy Levy/Marcella Detroit en
29991 Mikhail Shervashidze/Chachba en
31777 Philip Hammond/Hamman, Sr. hi
43349 Nanthakumar a/l Kaliappan ta
45738 Maruja Isabella Sevilla (Spanish Era), Nina Concepcion/Cristy (Reincarnation) es
46497 Victor a/l Andrag ro


In [149]:
replacechar('/', ' ', sol)

Lily Munster  OTS name: Phoebe Munster
Paula Baracho   Paula Ribeiro
Rajendra Choudhary   राजेन्द्रा चौधरी
Amanda Porter Emily Thorne
Anne-Marie Baiynd Presnell Buttrick Band
Líu Hóngdù (劉弘度)   Líu Bīn (劉玢)
Elavarasan a l Elangowan
Gay Stanhope Falcon Gay Laurence Tom Laurence Michael Mike Waring
Carsten Bo Eriksen   MBD73
Marcy Levy Marcella Detroit
Mikhail Shervashidze Chachba
Philip Hammond Hamman, Sr.
Nanthakumar a l Kaliappan
Maruja Isabella Sevilla (Spanish Era), Nina Concepcion Cristy (Reincarnation)
Victor a l Andrag


### Clean EXCLAMATION

In [150]:
exl = []
findname('EXCLAMATION', exl)

27704 Dr. Kucho! hi


In [151]:
replacechar('!', '', exl)

Dr. Kucho


### Clean DOLLAR

In [152]:
dol = []
findname('DOLLAR', dol)

35273 Ty Dolla $ign fi


In [153]:
replacechar('$', '', dol)

Ty Dolla ign


### Clean DAGGER

In [154]:
dag = []
findname('DAGGER', dag)

7182 Philipp zu Hohenfelden † de


In [155]:
replacechar('†', '', dag)

Philipp zu Hohenfelden 


### Clean LESS-THAN (drop)

In [156]:
lessth = []
findname('LESS-THAN', lessth)

44921 <span en


In [157]:
drop.append(lessth[0])
drop

[2816, 4319, 44921]

### Clean ACUTE (incomplete) (acute is an accent)

In [158]:
acu = []
findname('ACUTE', acu)

31033 Désiré M´Bonabucya fr


In [159]:
langname('M´Bonabucya')

['LATIN',
 'ACUTE',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

### Clean EQUALS

In [160]:
eq = []
findname('EQUALS', eq)

24351 =Giovanni Agostino Marliani it


In [161]:
replacechar('=', '', eq)

Giovanni Agostino Marliani


### Clean ZERO (ZERO WIDTH SPACE) (incomplete) (can't seem to get rid of it)

In [162]:
ze= []
findname('ZERO', ze)

13387 صادق قطب‌زاده fa


In [163]:
replacechar('​', '', ze)

صادق قطب‌زاده


### Clean APOSTROPHE (complete?) (only cleaned ones that seemed weird)

In [164]:
apo = []
findname('APOSTROPHE', apo)

147 Lawrence O'Brien en
263 William G. T'Vault en
553 John O'Donoghue en
600 Brendan O'Brien en
686 Breon O'Casey en
882 Detta O'Cathain gd
928 Junior Sa'u hr
1192 Ak'ak'i Meipariani ht
1248 Gary De'Roux fr
1419 Patrick O'Sullivan fi
1548 Sameh El-Saharty'' su
1949 Harriet Osborne O'Hagan en
2140 Cinders O'Brien en
2195 George 'Jocka' Todd en
2461 Rod O'Connor en
2570 Luke O'Connor en
2699 Mononc' Serge fr
2761 (Dato') it
2871 Aodh Ó'Raghallaigh ga
2972 Abdullah Sallum al-Samarra'i ar
2977 Abdillahi Suldaan Mohammed 'Timacade' so
3097 Graba' ro
3230 Steve O'Neill en
3339 Eleni Koiosi' fi
3839 Brittany O'Connell en
4006 Willem 's Gravesande en
4125 Paul O'Grady en
4184 Mark O'Brien en
4208 Anne d'Alençon fr
4297 Harvey O'Brien en
4342 Summer O'Brien en
4343 Martin Patrick O'Connell en
4367 F'Murr en
4405 Molly O'Day en
4420 Bryant O'Dare Hammett, Jr. en
4724 Ann O'Connell en
4746 सुर्नेद्र झा 'सुमन ' hi
4751 Conor Cruise O'Brien en
4782 Joseph Leonard O'Brien en
4887 John O'Neill en
492

In [165]:
replacechar("'''", '', apo)

Lawrence O'Brien
William G. T'Vault
John O'Donoghue
Brendan O'Brien
Breon O'Casey
Detta O'Cathain
Junior Sa'u
Ak'ak'i Meipariani
Gary De'Roux
Patrick O'Sullivan
Sameh El-Saharty''
Harriet Osborne O'Hagan
Cinders O'Brien
George 'Jocka' Todd
Rod O'Connor
Luke O'Connor
Mononc' Serge
(Dato')
Aodh Ó'Raghallaigh
Abdullah Sallum al-Samarra'i
Abdillahi Suldaan Mohammed 'Timacade'
Graba'
Steve O'Neill
Eleni Koiosi'
Brittany O'Connell
Willem 's Gravesande
Paul O'Grady
Mark O'Brien
Anne d'Alençon
Harvey O'Brien
Summer O'Brien
Martin Patrick O'Connell
F'Murr
Molly O'Day
Bryant O'Dare Hammett, Jr.
Ann O'Connell
सुर्नेद्र झा 'सुमन '
Conor Cruise O'Brien
Joseph Leonard O'Brien
John O'Neill
Anil Kumar Eravathri 
Joe O'Toole
Stephen James O'Donnell
Ed O'Bannon
Walter 'Dobie' Moore
Cassy O'Connor
J. J. O'Brien
Chak Tok Ich'aak II
Bill O'Brien
Chak Tok Ich'aak I
Leonello d'Este
Fawzi Abdel A'al
Tenema N'Diaye
April O'Neil
Tim O'Brien
Joseph O'Neill
Ta'Shia Phillips
Junior M'Pia Mapuku
Saia Fainga'a
Georg

### Clean LOW

In [166]:
low = []
findname('LOW', low)

8271 Panpayak_Jitmuangnon jw


In [167]:
replacechar('_', ' ', low)

Panpayak Jitmuangnon


### Clean FEMININE (incomplete) (unknown if should keep or not since its a unique character that can be identified to a languge)

In [168]:
fem = []
findname('FEMININE', fem)

31035 José Mª. Cabral Bermúdez es


### Clean MODIFIER (incomplete) (also seem to be unique characters to a language)

In [169]:
mod = []
findname('MODIFIER', mod)

19200 al-Ḥasan ibn ʻAlī al-Barbahārī ar
35526 Mulielealiʻi haw
35885 Israel Iz Kamakawiwoʻole haw
37882 Wikolia Kamehamalu Keawenui Kaahumanu-a-Kekūanaōʻa, Victoria Kamāmalu Kaahumanu IV haw
41238 Abū al-Ḥassan ʿAlī ibn ʿUthmān al-Marīni ar


In [170]:
langname('ʻAlī al-Barbahārī')

['MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'HYPHEN-MINUS',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

In [171]:
langname('ʿAlī ibn ʿUthmān al-Marīni')

['MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'HYPHEN-MINUS',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

### Clean RIGHT (312) (right parenthesis and right single quotation mark) and LEFT (301)

In [172]:
langname('’')


['RIGHT']

In [173]:
rig = []
lef = []
findname('RIGHT', rig)
findname('LEFT', lef)

46 David John (Dave) Shannon en
101 (Armen Lubin) de
197 C.W. (Bill) Hutchins en
355 Koshinaka Makoto (越中睦士) ja
555 Josip Seissel (Jo Klek) sl
762 Ancestral name: Jiang (姜) gu
802 Frank J. O’Connell en
915 (최영일) ko
1357 (Henry Roger) Justin Lewis en
1724 Hubert Duck, Deuteronomy Duck and Louis Duck (Quack Pack) en
1757 Charles D. “Chuck” McAtee en
1794 (GOSEComCComCS) en
1997 Junsei Terasawa (寺沢潤世) ja
2335 () en
2346 Pavel (Pavlo) Gintov sk
2411 (Τίνα Μπιρμπίλη) el
2446 Blessed John Finch (martyr) en
2479 David Hirst (arachnologist) en
2502 Hormisdas Etienne Djibri (Estefan Jabri) fr
2505 Baldwin (II) Rátót hu
2605 Kim Tae-Jin (김태진) ko
2608 (Serdar-ı Ekrem Ömer Paşa) tr
2761 (Dato') it
2942 Claire Kirby (Claire Debrois) en
2980 Saint Honoratus (Honorius) of Amiens la
3187 Milo Gunderson (body) en
3654 Alexander (Oleksandr) Khvostenko-Khvostov ru
4084 Kanysh Satpayev (Kanysh Satbaev) ru
4216 Mohammad Boroujerdi (محمد بروجردی) ar
4301 Örüg Temür Khan (Gulichi) tr
4308 Samuel Robles (Pana

In [174]:
replacechar(')', '', rig)

David John (Dave Shannon
(Armen Lubin
C.W. (Bill Hutchins
Koshinaka Makoto (越中睦士
Josip Seissel (Jo Klek
Ancestral name: Jiang (姜
Frank J. O’Connell
(최영일
(Henry Roger Justin Lewis
Hubert Duck, Deuteronomy Duck and Louis Duck (Quack Pack
Charles D. “Chuck” McAtee
(GOSEComCComCS
Junsei Terasawa (寺沢潤世
(
Pavel (Pavlo Gintov
(Τίνα Μπιρμπίλη
Blessed John Finch (martyr
David Hirst (arachnologist
Hormisdas Etienne Djibri (Estefan Jabri
Baldwin (II Rátót
Kim Tae-Jin (김태진
(Serdar-ı Ekrem Ömer Paşa
(Dato'
Claire Kirby (Claire Debrois
Saint Honoratus (Honorius of Amiens
Milo Gunderson (body
Alexander (Oleksandr Khvostenko-Khvostov
Kanysh Satpayev (Kanysh Satbaev
Mohammad Boroujerdi (محمد بروجردی
Örüg Temür Khan (Gulichi
Samuel Robles (Panamanian Musician
Stan Bowsher (footballer
Chen Tsu-Li (陳 祖烈
Frederick Charles (Karl, Duke of Württemberg-Winnental
Avrahm (Abraham Yarmolinsky
(The Right Honourable
Jeane (Jean Manson
Ardeshir (or in higher Persian: Ardashir Kamkar
(Italian Name Giuseppe Stefano Ch

In [175]:
checkempty = []
checkempty.append(replacechar('(', '', lef))

David John Dave Shannon
Armen Lubin
C.W. Bill Hutchins
Koshinaka Makoto 越中睦士
Josip Seissel Jo Klek
Ancestral name: Jiang 姜
최영일
Henry Roger Justin Lewis
Hubert Duck, Deuteronomy Duck and Louis Duck Quack Pack
Charles D. “Chuck” McAtee
GOSEComCComCS
Junsei Terasawa 寺沢潤世

Pavel Pavlo Gintov
Τίνα Μπιρμπίλη
Blessed John Finch martyr
David Hirst arachnologist
Hormisdas Etienne Djibri Estefan Jabri
Baldwin II Rátót
Kim Tae-Jin 김태진
Serdar-ı Ekrem Ömer Paşa
Dato'
Claire Kirby Claire Debrois
Saint Honoratus Honorius of Amiens
Milo Gunderson body
Alexander Oleksandr Khvostenko-Khvostov
Kanysh Satpayev Kanysh Satbaev
Mohammad Boroujerdi محمد بروجردی
Örüg Temür Khan Gulichi
Samuel Robles Panamanian Musician
Stan Bowsher footballer
Chen Tsu-Li 陳 祖烈
Frederick Charles Karl, Duke of Württemberg-Winnental
Avrahm Abraham Yarmolinsky
The Right Honourable
Jeane Jean Manson
Ardeshir or in higher Persian: Ardashir Kamkar
Italian Name Giuseppe Stefano Chiaffredo Marello English Joseph Stephen Jeffrey Marello


### Clean AMPERSAND

In [176]:
amp = []
findname('AMPERSAND', amp)

4309 Isaac & Miria en
22717 Tom & Olly en
24908 Saints Hor, Besoy, & Daydara bn
35732 Steven K. D&#39Arcy en
38598 Lanfranchi & Farina it


In [177]:
replacechar('&', '', amp)

Isaac  Miria
Tom  Olly
Saints Hor, Besoy,  Daydara
Steven K. D#39Arcy
Lanfranchi  Farina


### Clean WHITE

In [178]:
whi = []
findname('WHITE', whi)

19465 Beauty☆Takaco en


In [179]:
replacechar('☆', ' ', whi)

Beauty Takaco


### Clean MIDDLE

In [180]:
mid = []
findname('MIDDLE', mid)

20020 額爾德特·文繡 zh-CN
47690 愛新覺羅·胤禟 zh-CN


In [181]:
replacechar('·', ' ', mid)

額爾德特 文繡
愛新覺羅 胤禟


### Clean DIGIT (incomplete) (a little more complex)

In [182]:
dig = []
findname('DIGIT', dig)

195 Sir William Jardine, 7th Baronet en
1024 Auriane Mallo in 2013 la
2519 Coi4 Si1 Bui4 en
2766 Lok6 Dai3 ar
2816 thumb|right|Spitfire Mk IIa P7350 of the BBMF is the only existing airworthy Spitfire that fought in the Battle of Britain. en
2940 Ckay1 vi
3625 26th Clan Chief en
4112 Sir Edward Malet, 4th Bt en
4500 D-Ray 3000 en
4740 Robert de Holland, 1st Baron Holand es
6796 James Younger, 5th Viscount Younger of Leckie en
7745 Kyoko Inoue#5 ja
9952 Mateusz Rutkowski20px pl
12404 Caesar Lucius Aurelius Commodus Antoninus Augustus 180 la
12412 Can4 Jin4-hei1 en
12480 The Chain Gang of 1974 en
12817 William Pakenham, 4th Earl of Longford en
13902 George Douglas, 1st Earl of Dumbarton en
14063 9ice ar
14222 21 Savage en
14966 György Berencsi 3rd hu
15526 J57 ar
15874 1316-Sv.4 en
15914 Sir Charles Hastings, 1st Baronet en
16286 Gu2 Geoi6-gei1 bg
16428 Dik6 Lung4 en
17228 A1 ig
17400 Sir George White, 1st Baronet en
17700 wai6 laan4 ar
17701 1978 en
17727 Alexander Hood, 1st Viscount Br

18734 Heoi2 Gun3git6 ar
19889 John H. Ware 3rd en
21302 Ling4 Bo1 no
21891 John de Mowbray, 4th Baron Mowbray en
21965 Richard Arden, 3rd Baron Alvanley en
23111 Alacrana Plata #2 es
23549 Carsten Bo Eriksen   MBD73 da
23745 Sir Basil Samuel Hill-Wood, 2nd Baronet en
24342 Victor Durbal 03408.JPG bn
26310 1317-P.80 en
26403 William Willoughby, 5th Baron Willoughby de Eresby en
27846 Bennett Friedman age 67 en
28350 Bestia 666 es
28503 Sir Mordaunt Martin, 4th Baronet en
28648 Sir Francis Samuel Drake, 1st Baronet en
29121 John Neville, 4th Baron Latimer en
29382 William Brooke, 10th Baron Cobham en
29737 John de Mowbray, 1st Earl of Nottingham en
29759 1314-P.39 en
32216 2Mex en
32568 AK-69 ht
32719 Malise mac Gilleain, 2nd Chief gd
33174 12th Planet en
33522 C-3PO en
33732 1315 1899P.-2 en
34508 Sir William Cusack Smith, 2nd Baronet en
34963 Matthew Aylmer, 1st Baron Aylmer en
35044 1312-Sv.6 en
35045 E-40 en
35599 Masked Burning #2 en
35732 Steven K. D#39Arcy en
36673 Sir John Lafore

In [183]:
for i, x in enumerate(dig):
    if langname(df['name'][x][0]) == ['DIGIT']:
        print(x, df['name'][x], df['lang'][x])

3625 26th Clan Chief en
14063 9ice ar
14222 21 Savage en
15874 1316-Sv.4 en
17701 1978 en
26310 1317-P.80 en
29759 1314-P.39 en
32216 2Mex en
33174 12th Planet en
33732 1315 1899P.-2 en
35044 1312-Sv.6 en
39681 2Tap ar
46347 1321-P.44 en
46661 198 to 211 en
47230 1318-Ağ. Top1 az


### Clean EN (cleaned one name)

In [184]:
en = []
findname('EN', en)

14408 Stevan Nedić–Ćela hr
18503 Jerzy Waldorff–Preyss en
19030 Meinhard II of Gorizia–Tyrol en
38917 James Stirling 1800–1876 en


In [185]:
replacechar('1800–1876', '', en)

Stevan Nedić–Ćela
Jerzy Waldorff–Preyss
Meinhard II of Gorizia–Tyrol
James Stirling 


### Clean NUMBER (names with # may be duplicates)

In [186]:
num = []
findname('NUMBER', num)

7745 Kyoko Inoue#5 ja
23111 Alacrana Plata #2 es
35599 Masked Burning #2 en
35732 Steven K. D#39Arcy en
38372 James Gunnar Nixon #1 en


In [187]:
replacechar('#', '', num)

Kyoko Inoue5
Alacrana Plata 2
Masked Burning 2
Steven K. D39Arcy
James Gunnar Nixon 1


In [188]:
alphatype1 = []
for i in df['name']:
    for x in langname(i):
        alphatype1.append(x)

print(len(set(alphatype1)))

total1 = 0
for i in set(alphatype1):
    count = 0
    total1 +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total1)

41
418 COMMA
1 ZERO
2 TAMIL
3 ARMENIAN
1 KATAKANA
23 HANGUL
75 CYRILLIC
2 GEORGIAN
4 KANNADA
5 MODIFIER
31 COLON
77 ARABIC
13 HEBREW
83 DIGIT
2 HIRAGANA
1 SINHALA
1 LESS-THAN
1 MALAYALAM
5 BENGALI
47392 LATIN
2 IDEOGRAPHIC
1 TELUGU
9 LEFT
184 CJK
2489 FULL
1 ACUTE
1 MYANMAR
2 THAI
1400 HYPHEN-MINUS
14 GREEK
5 COMBINING
1 FEMININE
1 GURMUKHI
334 APOSTROPHE
1 ETHIOPIC
17 DEVANAGARI
20 RIGHT
3 EN
1 SYRIAC
46505 SPACE
2 VERTICAL
41


### Doing basic cleaning again

In [189]:
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
47754          Clyde Donaldson      1   en   
47755          Terry Alexander      1   en   
47756             Neil Roebuck      1   en   
47757             Lyle Stewart      1   en   
47758  Thomas Colclough Watson      1   en   

                                                alphabet  
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
...                                                  ...  
47754  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
47755  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
47756  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
47757  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
47758  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  

[47759 rows x 4 columns]

In [190]:
print(np.any(df["name"].isna()))
print(np.any(df["name"] == ''))
print(np.any(df["name"] == ' '))
print(np.any(df["name"] == '  '))

#Need to drop alphabets because df.duplicated() doesn't work with lists
df = df.drop(columns = 'alphabet')
print(np.any(df.duplicated()))

False
True
True
False
True


In [191]:
df.replace('', np.nan, inplace=True)
df.replace(' ', np.nan, inplace=True)

In [192]:
df = df.dropna()
df = df.drop_duplicates()

In [193]:
df['alphabet'] = df['name'].apply(langname)

Drop certain rows

In [194]:
df.reset_index(drop=True, inplace=True)
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
47749          Clyde Donaldson      1   en   
47750          Terry Alexander      1   en   
47751             Neil Roebuck      1   en   
47752             Lyle Stewart      1   en   
47753  Thomas Colclough Watson      1   en   

                                                alphabet  
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
...                                                  ...  
47749  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
47750  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
47751  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
47752  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
47753  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  

[47754 rows x 4 columns]

## Feature Creation

### accents (incomplete)

In [195]:
from unidecode import unidecode
import unicodedata

text_with_accents = "Létérs wïth âccénts"
normalized_text = unidecode(text_with_accents)
print(normalized_text)

Leters with accents


In [196]:
import unicodedata
langname = lambda x : unicodedata.name(x[0]).split(' ')[0]

In [197]:
def identify_accents(text):
    accents = []
    for char in text:
        if unicodedata.normalize('NFD', char) != char:
            accents.append(char)
    return accents
allaccents= set()
for i in df['name']:
    allaccents.update(identify_accents(i))
print(allaccents)
print(len(allaccents))

{'ī', 'Ḥ', '재', '송', 'è', 'Ó', 'ῑ', 'İ', 'ư', 'ǒ', 'Í', 'ō', 'ė', 'ো', '강', 'ạ', 'Ä', 'ć', 'ń', 'Ž', 'ọ', 'ş', 'ố', '윤', 'ô', 'š', 'ē', 'ì', 'ţ', '김', 'ŭ', 'Ḫ', 'Č', '상', 'Å', 'ě', 'ḫ', '철', 'ί', '대', 'â', 'Š', '권', 'Ľ', 'Â', 'й', 'ű', '일', 'ắ', '석', 'ğ', '노', 'ž', 'ę', '정', '하', '명', '연', '춘', 'č', '영', 'Ѓ', 'ứ', 'ụ', '건', 'ď', 'ț', 'ό', 'ň', 'ç', 'ೀ', 'ṭ', 'ü', 'ú', '신', 'ů', 'Ō', 'Ș', 'ǐ', 'ῖ', 'ä', 'ṣ', 'Ż', 'ź', '박', '헌', '랑', 'ò', 'ệ', '홍', '긍', 'أ', 'ș', 'ǫ', 'ǎ', 'î', '원', '태', '조', 'ả', 'ỹ', 'ģ', '형', 'ö', 'ņ', 'ợ', 'Ṭ', 'إ', '환', 'ê', '기', 'Ī', 'ấ', '민', 'ώ', 'ầ', 'ķ', 'ś', 'Ř', 'É', 'ؤ', 'Ț', 'ằ', 'ủ', 'ă', 'ơ', 'Ü', '광', '이', 'ớ', '진', 'ą', 'ờ', 'ữ', 'ế', '허', 'ǔ', '현', 'ά', '규', 'ā', 'ï', 'ĩ', 'ľ', 'ї', 'é', '최', 'í', '보', 'ñ', 'à', 'ř', 'Ć', '수', 'Ś', 'ೇ', 'õ', 'ÿ', 'á', 'Á', '섭', 'ৌ', 'ド', 'Ö', 'Ş', 'ã', '순', 'Ú', 'ề', 'ļ', 'ǚ', 'ù', 'ị', 'ý', 'ż', 'ồ', 'ő', 'å', 'Ç', 'ū', 'έ', 'ё', 'ũ', 'ë', 'ύ', 'û', 'ペ', '겸', 'ễ', 'ừ', '동', '녕', 'ó', 'ặ', '식'}
195


In [198]:
allaccents1 = []
for i in allaccents:
    if langname(str(i)) != 'HANGUL':
        allaccents1.append(i)
print(allaccents1)
print(len(allaccents1))


['ī', 'Ḥ', 'è', 'Ó', 'ῑ', 'İ', 'ư', 'ǒ', 'Í', 'ō', 'ė', 'ো', 'ạ', 'Ä', 'ć', 'ń', 'Ž', 'ọ', 'ş', 'ố', 'ô', 'š', 'ē', 'ì', 'ţ', 'ŭ', 'Ḫ', 'Č', 'Å', 'ě', 'ḫ', 'ί', 'â', 'Š', 'Ľ', 'Â', 'й', 'ű', 'ắ', 'ğ', 'ž', 'ę', 'č', 'Ѓ', 'ứ', 'ụ', 'ď', 'ț', 'ό', 'ň', 'ç', 'ೀ', 'ṭ', 'ü', 'ú', 'ů', 'Ō', 'Ș', 'ǐ', 'ῖ', 'ä', 'ṣ', 'Ż', 'ź', 'ò', 'ệ', 'أ', 'ș', 'ǫ', 'ǎ', 'î', 'ả', 'ỹ', 'ģ', 'ö', 'ņ', 'ợ', 'Ṭ', 'إ', 'ê', 'Ī', 'ấ', 'ώ', 'ầ', 'ķ', 'ś', 'Ř', 'É', 'ؤ', 'Ț', 'ằ', 'ủ', 'ă', 'ơ', 'Ü', 'ớ', 'ą', 'ờ', 'ữ', 'ế', 'ǔ', 'ά', 'ā', 'ï', 'ĩ', 'ľ', 'ї', 'é', 'í', 'ñ', 'à', 'ř', 'Ć', 'Ś', 'ೇ', 'õ', 'ÿ', 'á', 'Á', 'ৌ', 'ド', 'Ö', 'Ş', 'ã', 'Ú', 'ề', 'ļ', 'ǚ', 'ù', 'ị', 'ý', 'ż', 'ồ', 'ő', 'å', 'Ç', 'ū', 'έ', 'ё', 'ũ', 'ë', 'ύ', 'û', 'ペ', 'ễ', 'ừ', 'ó', 'ặ']
148


In [199]:
unicodedata.normalize('NFD', '안녕하세요') == '안녕하세요'

False

In [200]:
'안녕하세요' == '안녕하세요'

True

### alphabet

Use langname to find the alphabet name for each character in a name instead of just the first letter (Anna said its better to have full list) and create column

In [201]:
langname = lambda x:[unicodedata.name(char).split(' ')[0] for char in x]
df['alphabet'] = df['name'].apply(langname)
print(df['alphabet'].head())

0    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
1    [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...
2    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
3    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
4    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
Name: alphabet, dtype: object


In [202]:
allalphatype = []
#List of all alphabet types of every character in the dataset
for i in df['name']:
    for x in langname(i):
        allalphatype.append(x)

total = 0
for i in set(allalphatype):
    count = 0
    total +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total)

418 COMMA
1 ZERO
2 TAMIL
3 ARMENIAN
1 KATAKANA
23 HANGUL
75 CYRILLIC
2 GEORGIAN
4 KANNADA
5 MODIFIER
31 COLON
77 ARABIC
13 HEBREW
83 DIGIT
2 HIRAGANA
1 SINHALA
1 LESS-THAN
1 MALAYALAM
5 BENGALI
47390 LATIN
2 IDEOGRAPHIC
1 TELUGU
9 LEFT
184 CJK
2489 FULL
1 ACUTE
1 MYANMAR
2 THAI
1400 HYPHEN-MINUS
14 GREEK
5 COMBINING
1 FEMININE
1 GURMUKHI
334 APOSTROPHE
1 ETHIOPIC
17 DEVANAGARI
20 RIGHT
3 EN
1 SYRIAC
46504 SPACE
2 VERTICAL
41


Alphabets: (HEBREW, TELUGU, SINHALA, CJK, HIRAGANA, ARMENIAN, DEVANAGARI, KATAKANA, CYRILLIC, KANNADA,BENGALI, GURMUKHI, LATIN, TAMIL, GREEK, ARABIC, GEORGIAN, MALAYALAM, SYRIAC, HANGUL, MYANMAR, THAI, ETHIOPIC) (24)


Non alphabets: (ASTERISK, 418 COMMA, 2489 FULL, 1400 HYPHEN-MINUS, 5 COMBINING, 1 FULLWIDTH, 180 QUOTATION, 10 SEMICOLON, 4 VERTICAl, 31 COLON, 46500 SPACE, 15 SOLIDUS, 1 EXCLAMATION, 1 DOLLAR, 1 DAGGER, 1 LESS-THAN, 1 ACUTE, 1 EQUALS, 1 ZERO, 339 APOSTROPHE, 1 LOW, 2 IDEOGRAPHIC, 1 FEMININE, 5 MODIFIER, 312 RIGHT, 5 AMPERSAND, 1 WHITE, 301 LEFT, 2 MIDDLE, 85 DIGIT, 4 EN, 5 NUMBER) (32)

Check for double spaces

Cleaned: ASTERISK (cleaned), FULLWIDTH (cleaned), QUOTATION, SEMICOLON, VERTICAL (disappeared), 

Keep:FULL (period), COMBINING (accent), 

Skipped: COMMA, QUOTATION, COLON, EN (not sure if should remove), HYPHEN-MINUS (not sure if should remove)

In [203]:
char = []
for x, y in enumerate(df['name']):
    if 'ASTERISK' in (langname(y)):
        print(x, y, df.iloc[x, 2])
        char.append(x)
char

[]

Clean ASTERISK

In [204]:
tmp = [6816, 25458]
for y in tmp:
    df.iloc[y, 0]=df.iloc[y, 0].replace('*', '')
    print(df.iloc[y, 0])

df.iloc[12044, 0]=df.iloc[12044, 0].replace('*', ' ')
print({df.iloc[i, 0] for i in char})


Tuanku Fauziah
Čestmír Kožíšek
set()


Clean FULLWIDTH

In [205]:
fullw = []
for x, y in enumerate(df['name']):
    if 'FULLWIDTH' in (langname(y)):
        print(x, y, df.iloc[x, 2])
        fullw.append(x)
fullw

[]

In [206]:
df.iloc[43794, 0]= "Thomas Prence or Thomas Prince"
print(df.iloc[43794, 0])

Thomas Prence or Thomas Prince


Clean QUOTATION

In [207]:
quo = []
for x, y in enumerate(df['name']):
    if 'QUOTATION' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        quo.append(x)

In [208]:
for y in quo:
    df.iloc[y, 0]=df.iloc[y, 0].replace('"', '')
    print(df.iloc[y, 0])

Clean SEMICOLON

In [209]:
semi = []
for x, y in enumerate(df['name']):
    if 'SEMICOLON' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        semi.append(x)


In [210]:
for y in semi:
    df.iloc[y, 0]=df.iloc[y, 0].replace(';', '')
    print(df.iloc[y, 0])

Clean VERTICAL (dropped irrelevant)

In [211]:
ver = []
for x, y in enumerate(df['name']):
    if 'VERTICAL' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        ver.append(x)

2815 thumb|right|Spitfire Mk IIa P7350 of the BBMF is the only existing airworthy Spitfire that fought in the Battle of Britain. en
4318 frameless|center en


In [212]:
drop = [2816, 4319]
df.iloc[14690, 0] = 'Veronika Sramaty'
df.iloc[22302, 0] = "Timmy O'Dowd"

df['name'][ver]

2815    thumb|right|Spitfire Mk IIa P7350 of the BBMF ...
4318                                     frameless|center
Name: name, dtype: object

Clean COLON (incomp) (considering dropping all names)

In [213]:
col = []
for x, y in enumerate(df['name']):
    if 'COLON' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        col.append(x)

762 Ancestral name: Jiang 姜 gu
3595 German:  Josef Franz Leopold en
4091 German: Henriette Maria Norberta en
5304 Ardeshir or in higher Persian: Ardashir Kamkar en
6335 Lily Munster  OTS name: Phoebe Munster en
7432 File:Hollidaycamo.jpg en
15050 Farid Aslani Persian : فرید اصلانی fa
18354 Family name: Hóu 侯 en
19396 Dutch: Gabriël Boudewijn Karel Maria nl
20728 Dutch: Albert Leopold Clemens Maria Meinrad de
21738 Sajid AliUrdu:ساجد  علی fa
22194 Given name: Unknown en
24150 Arabic:عبد العزيز بديع صالح السليمي ar
25805 Chinese: Duàn Sīpíng 段思平 zh-CN
25875 Given name: Dao 導 ja
27428 Given name: Zhoupu 州蒲 en
27494 Makhdoom Khusro Bakhtyar Seraiki:مخدوم  کسر  بختٻد fa
27499 Laqab: Qutb ad-Din shortly uz
27578 Spanish: María Luisa de Borbón y Sajonia es
28804 Maganti Murali Mohan  Born : Maganti Raja Babu  te
31307 Traditional Chinese:韓世忠 zh-CN
34231 Given name: Dawud en
37174 Nu:Tone ro
40518 German: Wilhelm Friedrich Karl Ernst Joachim Albrecht en
40987 Family name:Sima司馬 zh-CN
42218 Giv

Clean SOLIDUS

In [214]:
sol = []
for x, y in enumerate(df['name']):
    if 'SOLIDUS' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        sol.append(x)


In [215]:
for y in sol:
    df.iloc[y, 0]=df.iloc[y, 0].replace('/', ' ')
    print(df.iloc[y, 0])

Clean EXCLAMATION

In [216]:
exl = []
for x, y in enumerate(df['name']):
    if 'EXCLAMATION' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        exl.append(x)

In [217]:
for y in exl:
    df.iloc[y, 0]=df.iloc[y, 0].replace('!', '')
    print(df.iloc[y, 0])

Clean DOLLAR

In [218]:
dol = []

for x, y in enumerate(df['name']):
    if 'DOLLAR' in (langname(y)):
        print(x,y, df.iloc[x, 2])
        dol.append(x)

In [219]:
for y in dol:
    df.iloc[y, 0]=df.iloc[y, 0].replace('$', '')
    print(df.iloc[y, 0])

Clean DAGGER

In [220]:
def findname(character, indexes):
    for x, y in enumerate(df['name']):
        if str(character) in (langname(y)):
            indexes.append(x)
            print(x, y, df.iloc[x, 2])

In [221]:
dag = []
findname('DAGGER', dag)

Drop certain rows all at once

In [222]:
#df = df.drop()
#df.reset_index(drop=True, inplace=True)

### avg_token_length (complete)

In [223]:
def average_token_length(name):
    total_length = 0
    total_tokens = 0
    tokens = name.split()

    # Calculate the total length of tokens in the name
    token_length = sum(len(token) for token in tokens)

    # Total length (characters in each word) and total number of tokens (words in name)
    total_length += token_length
    total_tokens += len(tokens)
    # Calculate the average token length
    if total_tokens > 0:
        average_length = total_length / total_tokens
        return average_length
    else:
        return 0 
    
avg_token_length = []
for i in df['name']:
    avg_token_length.append(average_token_length(i))

In [224]:
df['avg_token_length'] = avg_token_length

### num_tokens (complete)

In [225]:
def token_len(name):
    total_tokens = 0
    tokens = name.split()
    # Total number of tokens (words in name)
    total_tokens += len(tokens)
    return total_tokens

In [226]:
token_length = []
for i in df['name']:
    token_length.append(token_len(i))
df['num_tokens'] = token_length

### transliteration

In [227]:
transliteration = []
for i in df['name']:
    transliteration.append(unidecode(i))
df['transliteration'] = transliteration

In [228]:
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
47749          Clyde Donaldson      1   en   
47750          Terry Alexander      1   en   
47751             Neil Roebuck      1   en   
47752             Lyle Stewart      1   en   
47753  Thomas Colclough Watson      1   en   

                                                alphabet  avg_token_length  \
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   
...                                                  ...               ...   
47749  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47750  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47751  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47752  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47753  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               7.0   

       num_tokens          transliteration  
0               2             Raha Etemadi  
1               2   Leena Peltonen-Palotie  
2               2              Luma Grothe  
3               2            Takuya Kakine  
4               2           Ouyang Zhenhua  
...           ...                      ...  
47749           2          Clyde Donaldson  
47750           2          Terry Alexander  
47751           2             Neil Roebuck  
47752           2             Lyle Stewart  
47753           3  Thomas Colclough Watson  

[47754 rows x 7 columns]

### period_freq, dash_freq, apostrophe_freq, space_freq

In [229]:
df['period_freq'] = df['name'].apply(lambda name: name.count('.'))
df['dash_freq'] = df['name'].apply(lambda name: name.count('-'))
df['apostrophe_freq'] = df['name'].apply(lambda name: name.count('\''))
df['space_freq'] = df['name'].apply(lambda name: name.count(' '))

In [230]:
df.head()

name  class lang  \
0            Raha Etemadi      1   et   
1  Leena Peltonen-Palotie      1   fi   
2             Luma Grothe      1   en   
3           Takuya Kakine      1   ja   
4          Ōuyáng Zhènhuá      1   mi   

                                            alphabet  avg_token_length  \
0  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   

   num_tokens         transliteration  period_freq  dash_freq  \
0           2            Raha Etemadi            0          0   
1           2  Leena Peltonen-Palotie            0          1   
2           2             Luma Grothe            0          0   
3           2           Takuya Kakine            0          0   
4           2          Ouyang Zhenhua            0          0   

   apostrophe_freq  space_freq  
0                0           1  
1                0           1  
2                0           1  
3                0           1  
4                0           1

### Char n grams

In [231]:
import nltk
from nltk.util import ngrams
def generate_char_ngrams(text):
    if isinstance(text, str):
        words = text.split()
        unigrams = list(ngrams(text, 1))
        bigrams = list(ngrams(text, 2))
        trigrams = list(ngrams(text, 3))
        
        interpolated_ngrams = unigrams + bigrams + trigrams
        return interpolated_ngrams
    else:
        return []

### Word _ngrams

In [232]:
df['word_ngrams'] = df['name'].apply(lambda name: name.split())
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
47749          Clyde Donaldson      1   en   
47750          Terry Alexander      1   en   
47751             Neil Roebuck      1   en   
47752             Lyle Stewart      1   en   
47753  Thomas Colclough Watson      1   en   

                                                alphabet  avg_token_length  \
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   
...                                                  ...               ...   
47749  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47750  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47751  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47752  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47753  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               7.0   

       num_tokens          transliteration  period_freq  dash_freq  \
0               2             Raha Etemadi            0          0   
1               2   Leena Peltonen-Palotie            0          1   
2               2              Luma Grothe            0          0   
3               2            Takuya Kakine            0          0   
4               2           Ouyang Zhenhua            0          0   
...           ...                      ...          ...        ...   
47749           2          Clyde Donaldson            0          0   
47750           2          Terry Alexander            0          0   
47751           2             Neil Roebuck            0          0   
47752           2             Lyle Stewart            0          0   
47753           3  Thomas Colclough Watson            0          0   

       apostrophe_freq  space_freq                  word_ngrams  
0                    0           1              [Raha, Etemadi]  
1                    0           1    [Leena, Peltonen-Palotie]  
2                    0           1               [Luma, Grothe]  
3                    0           1             [Takuya, Kakine]  
4                    0           1            [Ōuyáng, Zhènhuá]  
...                ...         ...                          ...  
47749                0           1           [Clyde, Donaldson]  
47750                0           1           [Terry, Alexander]  
47751                0           1              [Neil, Roebuck]  
47752                0           1              [Lyle, Stewart]  
47753                0           2  [Thomas, Colclough, Watson]  

[47754 rows x 12 columns]

### A hat frequency

In [233]:
#trying to figure out accent distribution
print(allaccents1)
print(len(allaccents1))

['ī', 'Ḥ', 'è', 'Ó', 'ῑ', 'İ', 'ư', 'ǒ', 'Í', 'ō', 'ė', 'ো', 'ạ', 'Ä', 'ć', 'ń', 'Ž', 'ọ', 'ş', 'ố', 'ô', 'š', 'ē', 'ì', 'ţ', 'ŭ', 'Ḫ', 'Č', 'Å', 'ě', 'ḫ', 'ί', 'â', 'Š', 'Ľ', 'Â', 'й', 'ű', 'ắ', 'ğ', 'ž', 'ę', 'č', 'Ѓ', 'ứ', 'ụ', 'ď', 'ț', 'ό', 'ň', 'ç', 'ೀ', 'ṭ', 'ü', 'ú', 'ů', 'Ō', 'Ș', 'ǐ', 'ῖ', 'ä', 'ṣ', 'Ż', 'ź', 'ò', 'ệ', 'أ', 'ș', 'ǫ', 'ǎ', 'î', 'ả', 'ỹ', 'ģ', 'ö', 'ņ', 'ợ', 'Ṭ', 'إ', 'ê', 'Ī', 'ấ', 'ώ', 'ầ', 'ķ', 'ś', 'Ř', 'É', 'ؤ', 'Ț', 'ằ', 'ủ', 'ă', 'ơ', 'Ü', 'ớ', 'ą', 'ờ', 'ữ', 'ế', 'ǔ', 'ά', 'ā', 'ï', 'ĩ', 'ľ', 'ї', 'é', 'í', 'ñ', 'à', 'ř', 'Ć', 'Ś', 'ೇ', 'õ', 'ÿ', 'á', 'Á', 'ৌ', 'ド', 'Ö', 'Ş', 'ã', 'Ú', 'ề', 'ļ', 'ǚ', 'ù', 'ị', 'ý', 'ż', 'ồ', 'ő', 'å', 'Ç', 'ū', 'έ', 'ё', 'ũ', 'ë', 'ύ', 'û', 'ペ', 'ễ', 'ừ', 'ó', 'ặ']
148


In [234]:
#new column with names in all lower case
df['name_lower'] = df['name'].str.lower()
df['char_ngrams'] = df['name_lower'].apply(generate_char_ngrams)
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
47749          Clyde Donaldson      1   en   
47750          Terry Alexander      1   en   
47751             Neil Roebuck      1   en   
47752             Lyle Stewart      1   en   
47753  Thomas Colclough Watson      1   en   

                                                alphabet  avg_token_length  \
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   
...                                                  ...               ...   
47749  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47750  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
47751  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47752  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
47753  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               7.0   

       num_tokens          transliteration  period_freq  dash_freq  \
0               2             Raha Etemadi            0          0   
1               2   Leena Peltonen-Palotie            0          1   
2               2              Luma Grothe            0          0   
3               2            Takuya Kakine            0          0   
4               2           Ouyang Zhenhua            0          0   
...           ...                      ...          ...        ...   
47749           2          Clyde Donaldson            0          0   
47750           2          Terry Alexander            0          0   
47751           2             Neil Roebuck            0          0   
47752           2             Lyle Stewart            0          0   
47753           3  Thomas Colclough Watson            0          0   

       apostrophe_freq  space_freq                  word_ngrams  \
0                    0           1              [Raha, Etemadi]   
1                    0           1    [Leena, Peltonen-Palotie]   
2                    0           1               [Luma, Grothe]   
3                    0           1             [Takuya, Kakine]   
4                    0           1            [Ōuyáng, Zhènhuá]   
...                ...         ...                          ...   
47749                0           1           [Clyde, Donaldson]   
47750                0           1           [Terry, Alexander]   
47751                0           1              [Neil, Roebuck]   
47752                0           1              [Lyle, Stewart]   
47753                0           2  [Thomas, Colclough, Watson]   

                    name_lower  \
0                 raha etemadi   
1       leena peltonen-palotie   
2                  luma grothe   
3                takuya kakine   
4               ōuyáng zhènhuá   
...                        ...   
47749          clyde donaldson   
47750          terry alexander   
47751             neil roebuck   
47752             lyle stewart   
47753  thomas colclough watson   

                                             char_ngrams  
0      [(r,), (a,), (h,), (a,), ( ,), (e,), (t,), (e,...  
1      [(l,), (e,), (e,), (n,), (a,), ( ,), (p,), (e,...  
2      [(l,), (u,), (m,), (a,), ( ,), (g,), (r,), (o,...  
3      [(t,), (a,), (k,), (u,), (y,), (a,), ( ,), (k,...  
4      [(ō,), (u,), (y,), (á,), (n,), (g,), ( ,), (z,...  
...                                                  ...  
47749  [(c,), (l,), (y,), (d,), (e,), ( ,), (d,), (o,...  
47750  [(

In [236]:
import re
accent_counts = {accent: 0 for accent in allaccents1}
def count_accents(text, accent_counts):
    for accent in accent_counts.keys():
        accent_counts[accent] += len(re.findall(accent, text))
    return accent_counts
df['name_lower'].apply(lambda x: count_accents(x, accent_counts))
print(accent_counts)

{'ī': 15, 'Ḥ': 0, 'è': 66, 'Ó': 0, 'ῑ': 1, 'İ': 0, 'ư': 15, 'ǒ': 2, 'Í': 0, 'ō': 132, 'ė': 10, 'ো': 1, 'ạ': 3, 'Ä': 0, 'ć': 530, 'ń': 42, 'Ž': 0, 'ọ': 2, 'ş': 59, 'ố': 3, 'ô': 31, 'š': 270, 'ē': 3, 'ì': 7, 'ţ': 10, 'ŭ': 1, 'Ḫ': 0, 'Č': 0, 'Å': 0, 'ě': 11, 'ḫ': 2, 'ί': 9, 'â': 27, 'Š': 0, 'Ľ': 0, 'Â': 0, 'й': 24, 'ű': 1, 'ắ': 1, 'ğ': 49, 'ž': 90, 'ę': 6, 'č': 168, 'Ѓ': 0, 'ứ': 2, 'ụ': 2, 'ď': 1, 'ț': 18, 'ό': 1, 'ň': 11, 'ç': 143, 'ೀ': 4, 'ṭ': 2, 'ü': 241, 'ú': 138, 'ů': 6, 'Ō': 0, 'Ș': 0, 'ǐ': 4, 'ῖ': 1, 'ä': 85, 'ṣ': 1, 'Ż': 0, 'ź': 4, 'ò': 8, 'ệ': 4, 'أ': 9, 'ș': 30, 'ǫ': 1, 'ǎ': 3, 'î': 13, 'ả': 6, 'ỹ': 1, 'ģ': 2, 'ö': 282, 'ņ': 7, 'ợ': 3, 'Ṭ': 0, 'إ': 3, 'ê': 23, 'Ī': 0, 'ấ': 2, 'ώ': 2, 'ầ': 9, 'ķ': 1, 'ś': 13, 'Ř': 0, 'É': 0, 'ؤ': 1, 'Ț': 0, 'ằ': 1, 'ủ': 1, 'ă': 100, 'ơ': 11, 'Ü': 0, 'ớ': 1, 'ą': 12, 'ờ': 2, 'ữ': 1, 'ế': 6, 'ǔ': 2, 'ά': 9, 'ā': 35, 'ï': 24, 'ĩ': 1, 'ľ': 3, 'ї': 1, 'é': 1284, 'í': 641, 'ñ': 133, 'à': 28, 'ř': 28, 'Ć': 0, 'Ś': 0, 'ೇ': 1, 'õ': 8, 'ÿ': 1, 'á': 1046, '

In [235]:
def create_character_frequency_hashmap(df, names_col):
    char_freqs = {}
    for name in df[names_col]:
        for char in name:
            if char not in char_freqs.keys():
                char_freqs[char] = 1
    return char_freqs

### Frequency Distribution